src/Build/Graph/ProjectGraph.cs

```diff
@@ -250,6 +252,84 @@ private void LoadGraph(string entryProjectFile, ProjectCollection projectCollect
             }
         }
 
+        // Using Tarjan's algorithm for detecting strongly connected components in a directed graph
+        // Reference: https://en.wikipedia.org/wiki/Strongly_connected_component
```

Comment:
Tarjan's might be overkill here unless we're doing something with _all_ subgraphs of strongly connected components, but it looks like we just detect and print _some_ cycle.

Not too familiar with the static graph work so it's not an issue if this is just being called once for a build, but something like using a simple tri-colored DFS and a stack for printing the cycle would result in less allocations and overhead, and still ```O(|V| + |E|)```.